tools/autograd/gen_variable_type.py

```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
-    # the {} characters.
-    name_regex = r'(^|\W){}($|\W)'
+    for arg in args:
+        if 'name' not in arg:
+            # some returned arguments do not have names
+            continue
+        name = arg['name']
+
+        def replace_sizes(m):
+            res = name + '_sizes'
+            saved.append({'name': res, 'type': 'IntList'})
+            return res
+
+        def replace_zeros(m):
+            r = name + '_info'
+            saved.append({'name': r, 'type': 'TypeAndSize'})
+            return r + '.zeros()'
+
+        def replace_size_n(m):
+            res = name + '_argsize_{}'.format(*m.groups())
+            saved.append({'name': res, 'type': 'int64_t'})
+            return res
+
+        def replace_to_arg_sizes(m):
+            res = name + '_argsizes_{}'.format(*m.groups())
+            saved.append({'name': res, 'type': 'IntList'})
+            return res
+
+        # replace self.sizes() with self_sizes
+        formula = re.sub(r'{}.sizes\(\)'.format(name), replace_sizes, formula)
+        # replace zeros_like(self) with self_info
+        formula = re.sub(r'zeros_like\({}\)'.format(name), replace_zeros, formula)
+        # replace self.size(2) with self_size_2
+        formula = re.sub(r'{}.size\((\w+)\)'.format(name), replace_size_n, formula)
+        # replace to_arg_sizes(self, 2) with self_argsizes_2
+        formula = re.sub(r'to_arg_sizes\({}, (\w+)\)'.format(name), replace_to_arg_sizes, formula)
+
+        if re.search(IDENT_REGEX.format(name), formula):
+            arg = copy.deepcopy(arg)
+            arg['type'] = arg['type'].replace('const ', '').replace(' &', '')
+            saved.append(arg)
+    return formula, saved
+
+
+def create_derivative(declaration, formula, output_indices, var_names):
+    returns = [r for r in declaration['returns'] if r.get('name') != 'self']
+    arguments = [arg for arg in declaration['arguments']]
+    if any(arg['name'] == 'inplace' for arg in arguments):
+        for arg in arguments:
+            if arg['name'] == 'input':
+                returns += [arg]
+        arguments = [arg for arg in arguments if arg['name'] != 'input']
+    formula, saved_inputs = saved_variables(formula, arguments)
+    formula, saved_outputs = saved_variables(formula, returns)
+
+    return {
+        'formula': formula,
+        'output_indices': output_indices,
+        'saved_inputs': saved_inputs,
+        'saved_outputs': saved_outputs,
+        'var_names': var_names,
+    }
+
+
+def create_autograd_function(name, derivatives, num_inputs, buffers=None):
+    return {
+        'name': name,
+        'op': to_camel_case(name) + 'Backward',
+        'num_inputs': num_inputs,
+        'derivatives': derivatives,
+        'buffers': [] if buffers is None else buffers,
+        'saved_inputs': all_saved_variables(derivatives, 'saved_inputs'),
+        'saved_outputs': all_saved_variables(derivatives, 'saved_outputs'),
+    }
```

Comment:
Eh I'm starting to hate that this data is so unstructured. We should use objects with well defined attributes in this file (sort of like `namedtuple`, but probably we want to keep it mutable)